### ISSUE TITLE
Obsolete error should be reported for container class of a collection expression builder class

### ISSUE THREAD

**Issue by Spolutrean** [dotnet/roslyn#70742]
**Version Used**: visual studio17.8.0 Preview 1.0, .net80

**Steps to Reproduce**:

```csharp
[CollectionBuilder(typeof(Container.MyCollectionBuilder), "Create")]
public struct MyCollection<T> : IEnumerable<T>
{
    IEnumerator<T> IEnumerable<T>.GetEnumerator() => default;
    IEnumerator IEnumerable.GetEnumerator() => default;
}

[Obsolete("Message", error: true)]
public class Container
{
    public class MyCollectionBuilder
    {
        public static MyCollection<T> Create<T>(ReadOnlySpan<T> items) => default;
    }
}

class Program
{
    static void Main()
    {
        MyCollection<string> x = [];
        MyCollection<int> y = [1, 2, 3];
    }
}
```

[Sharplab](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQGED2AbbEDGMAlpgHYAyRMECU2yAPgAIBMAjALABQTADAARM2AOiy4CxMsgDc3PoLYAWWTwFDhAJTiliAWwijMugA5E8CAMo0AbkXwoVAbTF5CJUgCE4ZgCY0AFDAAnsYQmGD+WDpQRKQ0wgCyQS4S7l6+NACUADT8AEToSLAQeZkAunIAzPzIMAhwhPxJKW5kADwAKgB8/CAKlZ1d3ADe3Pzj/ACSAKKkcPq0MJgIg/2DwgDiEDCz8zSwy/6Z/AC8PX5gUHDYMCoTU7sLBwj9m9uP+0sIR6fnEJfXW7cAC+3G4jgA8sBkDhthB/Hl9MhkFAAOYQfgsPK5GgIZZ9OpwCDlKqCFj8KIwGJxBAjMYTJjVVhNZI4VySTzebB+Wlce6jPn3BlMtgANhZLQ5q0K0Gog38GmgPghpGwQQsxigpFWVAgumQxzO/AuVxudwmoK4lrBPHJQgA7HTBeMhOKmIomtSjvTxgKhRNmmzUu0hLwegAPU78Rxlc3+wPiVra2IwHpBKOONi5Fi5Sqxn1ChPs9xtTDAABWEh6AC8o3EAO5HOP8S3WrikKBIzX2BSabR6AxYExmGhWBC2ezIJ33RwAQRgdSIwDg1AAqij0f4C/d54vl9QOlAEOiYMhRNgoMj+Ax+LuEEuVxBD8ftmeLITGje7w+D0eT2fJh0GhLnsbJtwmQCAAsaF1Hwo0uegIDA50hVnXBMHrBJASIYw8HguhkGJCoUMZGpoDwODmUlNIuR5b99wxPp1Hox8Cz9f1BGqaiyHSbkaBY6h/A6EIMWXDIEGE0JclDfh9BgSDMB8AA5TtiX4YZLQ4zj+EkjFeJ5XT1P4E9pBbcCXRFAQEm2BTlNUoyTLMlDWyAA==)

**Diagnostic Id**: CS0619

**Expected Behavior**: `[]` and `[1, 2, 3]` are highlighted with obsolete error

**Actual Behavior**: there are no obsolete error highlightings on a `[]` and `[1, 2, 3]`

**Comment by Rekkonnect** [comment]
I think that the obsoletion errors should only be reported on the direct usage of the obsoleted symbol. As in the SharpLab example you provided, there already is a correctly reported obsoletion error on the usage of the `Container` class, which is enough to break the compilation. If you only had it as a warning, you would then still be able to use the MyCollectionBuilder class that is not actually marked as obsolete itself, it is only contained within an obsolete class and does not actually interact with it.

Note that if you mark `MyCollectionBuilder` itself as obsolete, you will actually get the obsoletion errors. And this is sufficient, because you are speficially denoting that class as obsolete, and not inferring its obsoletion via its container class. You might be more interested into an Obsolete attribute that also marks all nested members as obsolete.

**Comment by Spolutrean** [comment]
@Rekkonnect 
>I think that the obsoletion errors should only be reported on the direct usage of the obsoleted symbol.
If the `MyCollectionBuilder` is marked there will be an error/warning, but there is no "direct" usage of it. We are using the `Create` method inside, as well as we using inner class `MyCollectionBuilder` with a `Container` case. It is not clear what is considered "direct use" in this case. And why, when using a method inside a class, it is considered "direct use" of a class, and when using a class inside a parent class, it is not considered as "direct using" of a parent class.

**Comment by Rekkonnect** [comment]
> And why, when using a method inside a class, it is considered "direct use" of a class

It is not. [Example](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQGED2AbbEDGMAlpgHYAyRMECU2yAPgAIBMAjALABQ3AblAgAEAD0EBeQaQgB3QegAUASgDc3YQDoAsktU8uTACyCAGvPSD8i7gG9ug+xa07uAXweDu3ANoB5YMhwIagBdbiYAZkFWOUFbLncIqKNtRVi7BxdXIA=)

> If the MyCollectionBuilder is marked there will be an error/warning, but there is no "direct" usage of it.

It is. The syntax simply hides it, but the underlying collection type is directly used.

Now this sounds contradictory if we consider the `new()` feature in [this example](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQGED2AbbEDGMAlpgHYAyRMECU2yAPgAIBMAjALABQ36ABAA8+AXj6kIAdwAUASgDc3AQDoAsrIU8uTNgDY+/dElgRZ3AN7c+VvkwDsYyeu4BfbtwDaAeWDIcEagC63EwAzDYs+nwWXNY2YUwALHxqMlGW1q5czkA). We only get errors on the type itself, but not on the `new()` expression. But there is a good reason for this behavior (as a user), and that is because we are target-typing to an obsolete type, so we do not need extra errors indicating that we are directly using that type.

As a user, I can live and would rather live without those errors. The parent class **is not related** to the nested class **unless explicitly specified** (through inheritance, or by having a constructor accepting a parameter of the parent class type, etc.). The nested class contained in the parent class could very well be for structural purposes. Should the parent class be obsoleted, the nested class is not necessarily obsoleted. If we wanted to keep the nested class but not the parent class, we would move that nested class elsewhere, either nesting it in a different class, or having it directly contained in a namespace.

**Comment by Spolutrean** [comment]
I think "direct" usage should work for all nested classes. Let's look at an [example](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQGED2AbbEDGMAlpgHYAyRMECU2yAPgAIBMAjALABQA2gPLBkOCNQC63JgGYABK2lZSMKEVI1pAb27TtsmXICSpVQi07NXHZd2yALNICyACgCUG01asA3KAmlRpALzSqgDu0gJCeNTo2FDIyC4A3O5WAL4p6VyZ3PyCwmISeizheVEQMXHI3OapQA). Here is the container class is marked obsolete, and therefore we don't show highlighting when we create an instance of the obsolete class inside `M`.

**Comment by Rekkonnect** [comment]
Good counterpoint, it seems like a bug to me. Or an edge case that was not considered to align with behavior outside the scope of allowing obsolete elements. Might need to reconsider how obsoletion is evaluated as a whole. My point however remains that I would prefer no more errors to appear, and in your last example the warning should appear.

**Comment by Spolutrean** [comment]
I think it is not possible to reconsider now such a basic think in an obsolete annotation, because of the backward compatibility. For example [this](https://github.com/dotnet/roslyn/issues/65042) obvious bug won't be fixed for the same reason. So I guess it will be more consistent if we accept that all parent classes have a "direct" usage on the called code, and start handle it as I suggested in a ticket.

**Comment by Rekkonnect** [comment]
It would be *very* funny if the bug got intentionally fixed with the reasoning "time passed, and you still hold onto obsolete members". But yes, I see where this comes from, it's a largely breaking change. Moving forward, I would like to see those get corrected though. Obsolete members are meant to eventually be completely removed after all.

**Comment by jaredpar** [comment]
This is not covered in the spec but looking at similar behaviors I _believe_ the behavior is correct. Essentially where are other times that the compiler translates syntax into method calls on values and does it consider `Obsolete` on types / members. 

Consider `foreach` as an example: 
```csharp
D d = new();
foreach (var e in d)
{
}

[Obsolete]
class D 
{
    public class E
    {
        public bool MoveNext() => false;
        public int Current => default;
    }
   
    public E GetEnumerator() => default!;
}
```

This gives a diagnostic on the declaration but not on the `foreach`. If you add an `[Obsolete]` to say `GetEnumerator` then a diagnostic is issued in `foreach`. That seems to match the behavior for collection expressions here (presuming we do give a diagonstic when the member has the attribute. 

@cston, @RikkiGibson for feedback

**Comment by Spolutrean** [comment]
If we consider your example with `foreach`, it turns out that there is no need to highlight the error if the BuilderType is `Obsolete`, only if the function itself is `Obsolete`. But now we do highlight [such](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQGED2AbbEDGMAlpgHYAyRMECU2yAPgAIBMAjALABQTADAARM2AOiy4CxMsgDc3PoLYAWWTwFDhAJTiliAWwijMugA5E8CAMo0AbkXwoV3bgG0xeQiVIAhOGYAmNAAUMACexhCYYIEAsiFuEp4+/jQAlAA0/ABE6EiwEJkpALpyAMz8yDAIcIT8sfEeZAA8ACoAfPwgCiUtrdwA3tz8Q/wAkgCipHD6tDCYCD1dPcIA4hAwE1M0sHOBKfwAvO0BYFBw2DAqw6Mb09sIXStrN1uzCLsHRxAnZxfcAL5OLjOADywGQODWEECmWiKGQUAA5vkMjQEHNOpU4BAiqVBCxanEcO5JN5fNgAgh+oNhkwykIAGwE+okhY5aDUHqBDTQPzA0jYEIWYxQUgLKgQXTIPaHfjHU7nFQAriA1gKADsVK4VwZgkUtSgRFIu2pQwGWquwzqRISTSEvHaAA8DvxnIVLhahlbxA1RYaYO0Qs7nGwMiwMiU3Sb+EqldxSFB9MhhfYFJptHoDFgTGYaFYELZ7MhNVdnABBGCVIjAODUACq8KRgSjV3Ller1GaUAQSJgyFE2CgyGQ/AY/FbCCrNYgne7az7FkxNVH48nHa7Pb7Ix0NBO9jSzeGW4AFjRxX5nSd6BB9+aPaXcJgAO7RH5EYx4C90ZDY4q3wRlb86Agc9VWZRIyQpFd2wgDpUygqcozND0hlpfgwLIJJyRoeDqECZowhg6tkgQfDwgyO1+H0GAj0wPwADkE2xfg+iVZD/34UiYMwilOOY/ge2kaMDxQuk2AEWFqNohj9D4gShL/GMgA) case :c

**Comment by cston** [comment]
@Spolutrean, thanks for reporting this issue.
 
It looks like the compiler may not be consistent with when and where obsolete diagnostics are reported. We’ve discussed this within the Roslyn team and decided that for new features we’ll use the following rule for reporting obsolete diagnostics:
 
> If the source text *explicitly mentions* a type or member decorated with an ObsoleteAttribute, a diagnostic is reported.
 
This rule will be used for collection expressions and for new features going forward, but not otherwise applied to existing features.
 
For collection expressions, this means diagnostics for an obsolete collection builder method (or explicitly referenced containing type) will be reported at the attribute where the method is referenced, but not at each collection expression that implicitly uses the obsolete builder method.

**Comment by jaredpar** [comment]
Closing as the current implementation matches our defined behavior going forward
